### PR TITLE
bugfix/ch47215/product-error-messaging

### PR DESCRIPTION
**Pull request by busticated**
## Description

Improves error messaging when `--product <id>` flag is set but `device` param is not an ID ([ch](https://app.clubhouse.io/particle/story/47215/unhelpful-error-message-when-using-device-name-and-product-flag))

## How to Test

_Note: requires you to have a Particle product with one or more devices_

Run the following commands and note the error message you see:

`particle cloud flash <device-name> <path-to-project.ino> --product <product-id>`
`particle variable get <device-name> <var-name> --product <product-id>`
`particle subscribe --device <device-name> --product <product-id>`
`particle function call <device-name> <fn-name> --product <product-id>`

Then run `particle --help` and confirm all functions which currently support the `--product` flag as well as the `product` commands themselves properly handle being called with a `<device-name>`

## Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [x] Problem and solution clearly stated
- [x] Tests have been provided
- [x] Docs have been updated
- [x] CI is passing

### CODE REVIEW

File: src/cmd/base.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
+const { errors: { usageError } } = require('../app/command-processor');
+const spinnerMixin = require('../lib/spinner-mixin');
+const UI = require('../lib/ui');
+
+const DEVICE_ID_PTN = /^[0-9a-f]{24}$/i;
```
Comment (monkbroc) [inline src/cmd/base.js:5]: Should we be more permissive and accept capital hex letter too? The API lowercases it before doing the database query.

Comment (busticated) [reply src/cmd/base.js:5]: >Should we be more permissive and accept capital hex letter too?

it accepts capitals (note the `/i` modifier) - i added an assertion to make that clearer 👉 https://github.com/particle-iot/particle-cli/pull/570/commits/1c3b12df6e1e9f725f73b581c63edfd9e7cdf573#diff-d3d434b0fb05ede651109216b9967a98R31

File: src/cmd/base.test.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+			await promise;
+		} catch (e){
+			error = e;
+		}
+
+		expect(error).to.be.an.instanceof(Error);
+		expect(error).to.have.property('message', 'test');
+	});
+
+	it('Show a product device name usage error', async () => {
```
[the remark this answers is not in the document]
Comment (busticated) [reply src/cmd/base.test.js]: ✅ in https://github.com/particle-iot/particle-cli/commit/5a9d43ff3519914d984f821e8c6dca524df22078

File: test/e2e/subscribe.e2e.js [hunk 3]
```diff
@@ -342,28 +267,17 @@ describe('Subscribe Commands [@device]', () => {
...
-		await delay(5000);
-		subprocess.cancel(); // CTRL-C
-
-		const { all, isCanceled } = await subprocess;
-		const [subscribe,,, ...events] = all.split('\n');
-
-		expect(subscribe).to.equal(`Subscribing to all events from product ${PRODUCT_01_ID} device ${PRODUCT_01_DEVICE_02_ID}'s stream`);
-		expect(events).to.have.lengthOf.at.least(2);
+		const { events, received: [msg], isCanceled } = await collectEventOutput(args);
+
```
[the remark this answers is not in the document]
Comment (busticated) [reply test/e2e/subscribe.e2e.js:275]: ✅ in https://github.com/particle-iot/particle-cli/commit/5a9d43ff3519914d984f821e8c6dca524df22078

File: test/e2e/subscribe.e2e.js [hunk 4]
```diff
@@ -407,7 +320,56 @@ describe('Subscribe Commands [@device]', () => {
...
+		if (!Array.isArray(results) || !results.length){
+			return;
+		}
+		return results
+			.filter(str => str.trim().startsWith('{'))
+			.map(data => JSON.parse(data));
+	}
+
+	// TODO (mirande): capture stdout and stderr independently
+	async function collectEventOutput(...args){
```
[the remark this answers is not in the document]
Comment (busticated) [reply test/e2e/subscribe.e2e.js]: ✅ in https://github.com/particle-iot/particle-cli/commit/5a9d43ff3519914d984f821e8c6dca524df22078

**Pull request merged**